### PR TITLE
add deprecation warnings for API changes

### LINKED ISSUE #1383: add deprecation warnings for API changes
Ideally we should push out deprecation warnings as soon as we know it's happening.  Related issues:

- [x] #1373 Timeseries (targeted for 0.17)
- [x] #1377 Quick selectors (target 1.0)
- [x] #782 flags (target 1.0)

### PR DESCRIPTION
**Pull request by orbeckst**
Fixes #1383

Changes made in this Pull Request:
- closes #1383
- included deprecations:
  - #1373 Timeseries (targeted for 0.17)
    Note that the deprecation for core.Timeseries will always show up;
    this is deliberate so that users WILL see it as it will be gone in
    the next release!
  - #1377 Quick selectors (target 1.0)
  - #782 flags (target 1.0)
- updated CHANGELOG

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

File: package/MDAnalysis/coordinates/memory.py [hunk 1]
```diff
@@ -106,6 +106,8 @@
     universe2 = mda.Universe(PSF, coordinates, format=MemoryReader, order='afc')
 
 
+.. _create-in-memory-trajectory-with-AnalysisFromFunction:
```
Comment (kain88-de) [inline package/MDAnalysis/coordinates/memory.py]: why was this added?

File: package/MDAnalysis/core/Timeseries.py [hunk 2]
```diff
@@ -61,9 +70,19 @@
...
+from numpy.lib.utils import deprecate
+
 from . import groups
 
+with warnings.catch_warnings():
+    warnings.simplefilter('always', DeprecationWarning)
+    warnings.warn(('The Timeseries module and TimeseriesCollection will be '
+                   'removed in the 0.17 release. See issue #1372 '
+                   'https://github.com/MDAnalysis/mdanalysis/issues/1373'),
+                  DeprecationWarning)
```
Comment (kain88-de) [inline package/MDAnalysis/core/Timeseries.py]: This leaves the warning levels of other modules like ipython untouched?

File: package/MDAnalysis/core/__init__.py [hunk 3]
```diff
@@ -30,7 +30,8 @@
 :class:`~MDAnalysis.core.groups.AtomGroup` and return another
 :class:`~MDAnalysis.core.groups.AtomGroup`.
 
-:mod:`~MDAnalysis.Timeseries` are a convenient way to analyse trajectories.
+:mod:`~MDAnalysis.core.Timeseries` are a convenient way to analyse
+trajectories.
```
Comment (kain88-de) [inline package/MDAnalysis/core/__init__.py]: Can you remove this due to deprecations already here? We should already point people to use the analysis frame work

File: package/MDAnalysis/core/groups.py [hunk 4]
```diff
@@ -1487,12 +1471,46 @@ class AtomGroup(GroupBase):
...
+       *Instant selectors* of AtomGroup will be removed in the 1.0 release.
+       See issue `#1377
+       <https://github.com/MDAnalysis/mdanalysis/issues/1377>`_ for
+       more details.
+
     """
     def __getitem__(self, item):
+        # DEPRECATED in 0.16.2
+        # REMOVE in 1.0
+        #
```
Comment (kain88-de) [inline package/MDAnalysis/core/groups.py]: shouldn't there be warnings issues here?

File: package/MDAnalysis/core/groups.py [hunk 5]
```diff
@@ -1487,12 +1471,46 @@ class AtomGroup(GroupBase):
     AtomGroup instances are always bound to a
     :class:`MDAnalysis.core.universe.Universe`. They cannot exist in isolation.
 
+
+    .. rubric:: Deprecated functionality
+
+    *Instant selectors* will be removed in the 1.0 release.  See issue `#1377
+    <https://github.com/MDAnalysis/mdanalysis/issues/1377>`_ for more details.
+
+    Atoms can also be accessed in a Pythonic fashion by using the atom name as
```
Comment (richardjgowers) [inline package/MDAnalysis/core/groups.py]: If we're deprecating something, can we remove the documentation for it?  Someone might skip the dep warning, then read about this great feature they want to use forevermore....

File: package/MDAnalysis/coordinates/memory.py (hunk 1 above)
Comment (orbeckst) [inline package/MDAnalysis/coordinates/memory.py]: So that I can reference the anchor in the docs in package/MDAnalysis/core/Timeseries.py – it can be removed again in 0.17.0 when Timeseries.py is gone. This is the standard way to do cross references in sphinx.

I do not like the idea of removing the docs (see below) but rather point to alternatives. This will be the last release with the feature. If you ever need to tell someone which release to use then you point them to 0.16.2. And then the docs for the feature should still exist.

File: package/MDAnalysis/core/Timeseries.py (hunk 2 above)
Comment (orbeckst) [inline package/MDAnalysis/core/Timeseries.py]: I presume so... it's what we have been doing elsewhere.

File: package/MDAnalysis/core/__init__.py (hunk 3 above)
Comment (orbeckst) [inline package/MDAnalysis/core/__init__.py]: Ok, will do.

File: package/MDAnalysis/core/groups.py (hunk 4 above)
Comment (orbeckst) [inline package/MDAnalysis/core/groups.py]: This will annoy the hell of people, won't it? I can't use the decorator because `__getitem__` is used for other stuff.

File: package/MDAnalysis/core/groups.py (hunk 5 above)
Comment (orbeckst) [inline package/MDAnalysis/core/groups.py]: I don't want to remove docs of a feature in the last release where the feature exists. If anyone really wants the feature then they should still have the docs. Deprecation warnings should be enough. And it's already in its own section.

File: package/MDAnalysis/core/groups.py (hunk 4 above)
Comment (orbeckst) [inline package/MDAnalysis/core/groups.py]: Ok, adding, the warnings only show up once.

**Comment by orbeckst** [comment]
@richardjgowers how do `ResidueGroup` and `SegmentGroup` do their instant selector magic? 
```python
import MDAnalysis as mda; from MDAnalysisTests.datafiles import PSF
u = mda.Universe(PSF)

print(u.residues.LYS)
print(u.segments.s4AKE)
```
I don't see a `__getattr__` in them or in `GroupBase` so I don't know how to attach deprecation warnings.

File: package/MDAnalysis/core/Timeseries.py (hunk 2 above)
Comment (orbeckst) [inline package/MDAnalysis/core/Timeseries.py]:
```
In [15]: warnings.catch_warnings?
Init signature: warnings.catch_warnings(self, record=False, module=None)
Docstring:
A context manager that copies and restores the warnings filter upon
exiting the context.
```

File: package/MDAnalysis/core/groups.py (hunk 4 above)
Comment (kain88-de) [inline package/MDAnalysis/core/groups.py]: Well we only have a warning when `item` is a string right?

File: package/MDAnalysis/core/Timeseries.py (hunk 2 above)
Comment (kain88-de) [inline package/MDAnalysis/core/Timeseries.py]: Good just checking

**Comment by kain88-de** [comment]
OK my pushes are now showing up in the PR's that is weird

**Comment by orbeckst** [comment]
Thanks, I will merge with what I am working on.

**Comment by richardjgowers** [comment]
@orbeckst 
 - [here for named atom](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/core/topologyattrs.py#L412)
 - [here for named residues](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/core/topologyattrs.py#L1060)
 - [here for named segment](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/core/topologyattrs.py#L1247)

You should be able to use the `@deprecate` decorator on those and it'll only appear if you use them

**Comment by orbeckst** [comment]
@richardjgowers how does this end up in `__getattr__`?

**Comment by orbeckst** [comment]
... or have a look at the PR and comment, that is probably going to be more helpful. Or push to the PR.

I have to do other things for the rest of the day, sorry.

**Comment by orbeckst** [comment]
Note: At least two of the new tests are currently failing because I didn't know how add depreactions:
```
MDAnalysisTests.core.test_groups.TestInstantSelectorDeprecationWarnings.test_AtomGroup_nowarn_getitem_index ... ok
MDAnalysisTests.core.test_groups.TestInstantSelectorDeprecationWarnings.test_AtomGroup_warn_getattr ... ok
MDAnalysisTests.core.test_groups.TestInstantSelectorDeprecationWarnings.test_AtomGroup_warn_getitem ... ok
MDAnalysisTests.core.test_groups.TestInstantSelectorDeprecationWarnings.test_ResidueGroup_warn_getattr_resname ... FAIL
MDAnalysisTests.core.test_groups.TestInstantSelectorDeprecationWarnings.test_SegmentGroup_nowarn_getitem ... ok
MDAnalysisTests.core.test_groups.TestInstantSelectorDeprecationWarnings.test_SegmentGroup_warn_getattr ... FAIL
MDAnalysisTests.core.test_groups.TestInstantSelectorDeprecationWarnings.test_Segment_warn_getattr_rRESNUM ... ok
MDAnalysisTests.core.test_groups.TestInstantSelectorDeprecationWarnings.test_Segment_warn_getattr_resname ... ok
```

File: package/MDAnalysis/core/groups.py (hunk 5 above)
Comment (orbeckst) [inline package/MDAnalysis/core/groups.py]: I confused the instant selectors with the Timeseries regarding "last release". But I still think that it should be documented for as long as it is available. 

(Otherwise the *deprecated:: 0.16.2* tag makes little sense: it says when it was deprecated, not when it was removed. The removal also kills the docs and possible leaves a *versionchanged:: 1.0.0* behind.)

**Comment by orbeckst** [comment]
I am trying https://github.com/MDAnalysis/mdanalysis/pull/1403#issuecomment-309136416.

**Comment by orbeckst** [comment]
With decorators on the methods in  https://github.com/MDAnalysis/mdanalysis/pull/1403#issuecomment-309136416 and explicit warnings in one other place (for the Segment.r1 selectors) it now seems to be working as expected, at least locally.

I will rebase and clean up.

**Comment by orbeckst** [comment]
Interactive work will be annoying for a while because TAB completion on ``u.residues.<TAB>` will trigger the deprecation warning. Depending on settings for the warning filter, this will only show up once, though, and then deprecation warnings for real use will be swallowed. Let's see how it goes... if it is too annoying we can fix it (... how? ...) as we go along.

**Comment by orbeckst** [comment]
Rebased and force-pushed.

If travis is green (any my local docs look ok) then this is good to go from my side.

**Comment by orbeckst** [comment]
sphinx fails, not sure why:
```
/Volumes/Data/oliver/Biop/Projects/Methods/MDAnalysis/mdanalysis/package/MDAnalysis/core/Timeseries.py:82: DeprecationWarning: The Timeseries module and TimeseriesCollection will be removed in the 0.17 release. See issue #1372 https://github.com/MDAnalysis/mdanalysis/issues/1373
  DeprecationWarning)
reading sources... [100%] index

Warning, treated as error:
/Volumes/Data/oliver/Biop/Projects/Methods/MDAnalysis/mdanalysis/package/MDAnalysis/coordinates/DCD.py:docstring of MDAnalysis.coordinates.DCD.DCDReader.correl:11: (SEVERE/4) Unexpected section title.

Parameters
----------
```

The only change is `@deprecate` decorator on DCD.correl.

**Comment by kain88-de** [comment]
I fixed all sphinx warnings.

**Comment by kain88-de** [comment]
If you use the numpy `deprecate` decorator sphinx might fail because the decorator changes the docstring. This seems to be problematic for class member functions.

**Comment by kain88-de** [comment]
Can someone else please have a look at the failing tests. I don't find where we trigger the warnings that fail for the old build.

File: package/MDAnalysis/core/topologyattrs.py [hunk 6]
```diff
@@ -409,6 +411,9 @@ def getattr__(atomgroup, name):
             raise AttributeError("'{0}' object has no attribute '{1}'".format(
                     atomgroup.__class__.__name__, name))
 
+    @deprecate(message="Instant selector AtomGroup['<name>'] or AtomGroup.<name> "
+               "is deprecated and will be removed in 1.0. "
+               "Use AtomGroup.select_atoms('name <name>') instead.")
     def _get_named_atom(group, name):
```
Comment (kain88-de) [inline package/MDAnalysis/core/topologyattrs.py]: We now issue a warning for `atoms.ChainIDs` or `atoms.segids`. I don't think that was the indented effect. @richardjgowers could you have a look please.

File: package/MDAnalysis/core/groups.py [hunk 7]
```diff
@@ -1513,6 +1513,9 @@ def __getitem__(self, item):
         #
         # u.atoms['HT1'] access, otherwise default
         if isinstance(item, string_types):
+            warnings.warn("Using the [] operator with strings is deprecated."
```
Comment (orbeckst) [inline package/MDAnalysis/core/groups.py]: @kain88-de thanks for adding it back – I must have removed it accidentally when I rebased.

**Comment by orbeckst** [comment]
@kain88-de thanks for working on the PR and figuring out the deprecation decorator/sphinx issue as well as putting back 13bc527.

File: package/MDAnalysis/core/topologyattrs.py (hunk 6 above)
Comment (orbeckst) [inline package/MDAnalysis/core/topologyattrs.py]: @kain88-de I added a test if AtomGroup.segids raises a warning and locally the test passes. When I try it in ipython then 

In python and ipython, no warning is raised for `atoms.segids`:
```python
Python 2.7.13 (default, Apr 25 2017, 03:00:22)
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import MDAnalysis as mda; from MDAnalysisTests.datafiles import PSF; import MDAnalysisTests
>>> w = MDAnalysisTests.make_Universe(("resids", "resnames", "segids", "names"))
>>> w
<Universe with 125 atoms>
>>> w.atoms.segids
array(['SegA', 'SegA', 'SegA', 'SegA', 'SegA', 'SegA', 'SegA', 'SegA',
       'SegA', 'SegA', 'SegA', 'SegA', 'SegA', 'SegA', 'SegA', 'SegA',
       'SegA', 'SegA', 'SegA', 'SegA', 'SegA', 'SegA', 'SegA', 'SegA',
       'SegA', 'SegB', 'SegB', 'SegB', 'SegB', 'SegB', 'SegB', 'SegB',
       'SegB', 'SegB', 'SegB', 'SegB', 'SegB', 'SegB', 'SegB', 'SegB',
       'SegB', 'SegB', 'SegB', 'SegB', 'SegB', 'SegB', 'SegB', 'SegB',
       'SegB', 'SegB', 'SegC', 'SegC', 'SegC', 'SegC', 'SegC', 'SegC',
       'SegC', 'SegC', 'SegC', 'SegC', 'SegC', 'SegC', 'SegC', 'SegC',
       'SegC', 'SegC', 'SegC', 'SegC', 'SegC', 'SegC', 'SegC', 'SegC',
       'SegC', 'SegC', 'SegC', 'SegD', 'SegD', 'SegD', 'SegD', 'SegD',
       'SegD', 'SegD', 'SegD', 'SegD', 'SegD', 'SegD', 'SegD', 'SegD',
       'SegD', 'SegD', 'SegD', 'SegD', 'SegD', 'SegD', 'SegD', 'SegD',
       'SegD', 'SegD', 'SegD', 'SegD', 'SegE', 'SegE', 'SegE', 'SegE',
       'SegE', 'SegE', 'SegE', 'SegE', 'SegE', 'SegE', 'SegE', 'SegE',
       'SegE', 'SegE', 'SegE', 'SegE', 'SegE', 'SegE', 'SegE', 'SegE',
       'SegE', 'SegE', 'SegE', 'SegE', 'SegE'], dtype=object)
>>>
>>> w.atoms['AAA']
/Volumes/Data/oliver/Biop/Projects/Methods/MDAnalysis/mdanalysis/package/MDAnalysis/core/groups.py:1520: DeprecationWarning: Using the [] operator with strings is deprecated and will be removed in 1.0. Please use `select_atoms('name AAA')` instead.
  category=DeprecationWarning)
/Volumes/Data/oliver/Biop/Projects/Methods/MDAnalysis/mdanalysis/package/MDAnalysis/core/groups.py:1522: DeprecationWarning: `_get_named_atom` is deprecated!
Instant selector AtomGroup['<name>'] or AtomGroup.<name> is deprecated and will be removed in 1.0. Use AtomGroup.select_atoms('name <name>') instead.
  return self._get_named_atom(item)
<Atom 1: AAA of resname RsA, resid 1 and segid SegA>
```

Could you please paste what code you've been running?

File: package/MDAnalysis/core/groups.py (hunk 4 above)
Comment (orbeckst) [inline package/MDAnalysis/core/groups.py]: done

**Comment by orbeckst** [comment]
https://github.com/MDAnalysis/mdanalysis/pull/1403#issuecomment-309203849

> Can someone else please have a look at the failing tests. I don't find where we trigger the warnings that fail for the old numpy version build.

It is weird that the OLD_NUMPY tests fail but the new ones do not – if anything, the new ones should fail, too! It seems that `_get_named_atom()` is not only used for instant selectors but also for other things, e.g., in the CDRWriter or PQRWriter. https://travis-ci.org/MDAnalysis/mdanalysis/jobs/244102138#L1310

What should we do? Create a copy of `_get_named_atom()` and deprecate one? But I would need @richardjgowers help in figuring out where this goes. I don't fully understand the new topology magic yet...

File: package/MDAnalysis/core/topologyattrs.py (hunk 6 above)
Comment (kain88-de) [inline package/MDAnalysis/core/topologyattrs.py]: I adopted `_get_named_atom` to also print the requested name. That way I found out we apparently use it for more then just atom names

```python
from MDAnalysisTests import make_Universe
import warnings
warnings.simplefilter('always')
req_attrs = {'resnames': 'UNK',
             'resids': 1,
             'names': 'X',
             'tempfactors': 0.0,
             }
attrs = list(req_attrs.keys())
u = make_Universe(attrs, trajectory=True)
u.atoms.write('test.crd')
```

output

```bash
/home/max/foss/molecular-dynamics/mdanalysis/mdanalysis/package/MDAnalysis/core/groups.py:1539: DeprecationWarning: `_get_named_atom` is deprecated!
Instant selector AtomGroup['<name>'] or AtomGroup.<name> is deprecated and will be removed in 1.0. Use AtomGroup.select_atoms('name <name>') instead.
  return self._get_named_atom(attr)
chainIDs
/home/max/foss/molecular-dynamics/mdanalysis/mdanalysis/package/MDAnalysis/core/groups.py:1539: DeprecationWarning: `_get_named_atom` is deprecated!
Instant selector AtomGroup['<name>'] or AtomGroup.<name> is deprecated and will be removed in 1.0. Use AtomGroup.select_atoms('name <name>') instead.
  return self._get_named_atom(attr)
segids
/home/max/foss/molecular-dynamics/mdanalysis/mdanalysis/package/MDAnalysis/coordinates/CRD.py:229: UserWarning: Supplied AtomGroup was missing the following attributes: tempfactors. These will be written with default values. 
  "".format(miss=', '.join(missing_topology)))
```

This is running with numpy 1.12.1

**Comment by kain88-de** [comment]
> It is weird that the OLD_NUMPY tests fail but the new ones do not

Maybe there is a bug in asserts warn in newer numpy versions? But the bug it detected is definitely real on our side.

File: testsuite/MDAnalysisTests/util.py [hunk 8]
```diff
@@ -148,3 +149,50 @@ def in_dir(dirname):
     os.chdir(dirname)
     yield dirname
     os.chdir(old_path)
+
+
+def assert_nowarns(warning_class, *args, **kwargs):
```
Comment (jbarnoud) [inline testsuite/MDAnalysisTests/util.py]: Is there a test for this function? I do not see it, but I may have missed it. Note however that @utkbansal may be working on `test_util.py` and asked for it to be frozen.

File: package/MDAnalysis/coordinates/DCD.py [hunk 9]
```diff
@@ -578,37 +579,39 @@ def timeseries(self, asel=None, start=None, stop=None, step=None, skip=None,
...
-             of the first frame in the trajectory); the default ``None`` starts
-             at the beginning.
-        stop : int (optional)
-             End reading the trajectory at frame index `stop`-1, i.e, `stop` is excluded.
-             The trajectory is read to the end with the default ``None``.
-        step : int (optional)
-             Step size for reading; the default ``None`` is equivalent to 1 and means to
-             read every frame.
+        """
+Populate a :class:`~MDAnalysis.core.Timeseries.TimeseriesCollection` object
```
Comment (kain88-de) [inline package/MDAnalysis/coordinates/DCD.py]: The decorator adds to the docstring and messes up sphinx.

Comment (jbarnoud) [inline package/MDAnalysis/coordinates/DCD.py]: Could there be a comment about that, then? I am ready to bet it is going to be reverted by accident.

Comment (kain88-de) [inline package/MDAnalysis/coordinates/DCD.py]: This is going to be removed in the next release anyway. I already have a PR open for it.

File: testsuite/MDAnalysisTests/util.py (hunk 8 above)
Comment (orbeckst) [inline testsuite/MDAnalysisTests/util.py]: Have we got tests for the functions that we define for our test suite?

It has nothing to do with `test_util.py`, which tests MDAnalysis.lib.util.

**Comment by orbeckst** [comment]
>> It is weird that the OLD_NUMPY tests fail but the new ones do not
> Maybe there is a bug in asserts warn in newer numpy versions? But the bug it detected is definitely real on our side.

It could also be that the warnings are in a different order. The assertion is triggered because the `DeprecationWarning` is the *first* warning (instead of `UserWarning`).

In any case, it doesn't seem to be as simple as "deprecate [_get_named_atom()](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/core/topologyattrs.py#L412)` because it is used in `[topologyattrs.Atomnames.getatrr__](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/core/topologyattrs.py#L405)` and that is transplanted "everwhere" (as is `_get_named_atom`). There is also a check in `_get_named_atom` to [either return a singlet or a group](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/core/topologyattrs.py#L426) – I am not sure if this should go as well.

In summary, without @richardjgower 's help or someone else who knows the new topology system really well I don't know where to put the deprecation warnings.

File: testsuite/MDAnalysisTests/util.py (hunk 8 above)
Comment (jbarnoud) [inline testsuite/MDAnalysisTests/util.py]: Sorry for the confusion. I was convinced we had tests for the test utils. It would be a good idea, but for an other time and PR.

**Comment by richardjgowers** [comment]
Ok, I think what we have to do is add the DepWarning in the successful resolution of `_get_named_atom` @kain88-de I think is looking at this.  The problem is that this method is called whenever we hit a `__getattr__` error, and we need to filter between nonexistant attributes (ie looking up `ag.chainIDs` when they don;t exist) and actual uses of the quick selector.

**Comment by kain88-de** [comment]
@richardjgowers if this solves the problem we might be able to just do a normal style release.

**Comment by kain88-de** [comment]
It seems to have worked. The old numpy test isn't failing anymore.

**Comment by kain88-de** [comment]
@richardjgowers do you want to merge this? 

@jbarnoud did you see anything else that needs to be fixed?

**Comment by richardjgowers** [comment]
I should probably clone it and play with it locally to check it.  Give me a few hours to testdrive it

**Comment by orbeckst** [comment]
The full test times out, everything else seems to be ok. @richardjgowers you have the honor of merging...
